### PR TITLE
Add trim support to zpool wait

### PR DESCRIPTION
**Pull request by jgallag88**
### Motivation and Context
https://github.com/openzfs/zfs/pull/9162 added `zpool wait`, which provides a mechanism for waiting synchronously for the completion of long-running pool activities such as scrubs and device removals. `zpool wait` doesn't support waiting for the completion of manual `trim` operations though, since it was written initially before `trim` support was integrated. However, manual trims fall into the category of long-running pool activities which people might want to wait synchronously for, so this change adds the missing trim support to `zpool wait`.

### Description
With this change, users can run `zpool wait -t trim <pool>` to wait for all of the vdevs in the pool with trims in progress to finish being trimmed. They can also add a `-w` flag to the command that starts the trim (i.e. `zpool trim -w <pool>`) to make `zpool trim` synchronous.

The logic to wait for trims to complete is implemented in a very similar manner to the logic to wait for device initialization to complete, since the interfaces for and implementations of trimming and initialization are very similar.

### How Has This Been Tested?
Added tests to cover the new functionality.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: cmd/zpool/zpool_main.c [hunk 1]
```diff
@@ -9466,21 +9479,30 @@ zpool_do_set(int argc, char **argv)
...
-vdev_initialize_remaining(nvlist_t *nv)
+vdev_activity_remaining(nvlist_t *nv, zpool_wait_activity_t activity)
 {
 	uint64_t bytes_remaining;
 	nvlist_t **child;
 	uint_t c, children;
 	vdev_stat_t *vs;
 
+	ASSERT(activity == ZPOOL_WAIT_INITIALIZE ||
+	    activity == ZPOOL_WAIT_TRIM);
```
Comment (behlendorf) [inline cmd/zpool/zpool_main.c]: Please use `assert` instead of `ASSERT` in the user space commands and libraries.

File: module/zfs/spa.c [hunk 2]
```diff
@@ -9280,28 +9280,35 @@ spa_wake_waiters(spa_t *spa)
 	mutex_exit(&spa->spa_activities_lock);
 }
 
-/* Whether the vdev or any of its descendants is initializing. */
+/* Whether the vdev or any of its descendants is being initialized/trimmed. */
```
Comment (behlendorf) [inline module/zfs/spa.c]: nit: "is being.." -> "are being.." sounds better to me

File: module/zfs/spa.c [hunk 3]
```diff
@@ -9280,28 +9280,35 @@ spa_wake_waiters(spa_t *spa)
...
+	    &vd->vdev_initialize_lock : &vd->vdev_trim_lock;
 
 	mutex_exit(&spa->spa_activities_lock);
-	mutex_enter(&vd->vdev_initialize_lock);
+	mutex_enter(lock);
 	mutex_enter(&spa->spa_activities_lock);
 
-	initializing = (vd->vdev_initialize_state == VDEV_INITIALIZE_ACTIVE);
-	mutex_exit(&vd->vdev_initialize_lock);
+	boolean_t in_progress = activity == ZPOOL_WAIT_INITIALIZE ?
```
Comment (behlendorf) [inline module/zfs/spa.c]: nit: Can you add `( )` around the conditional here.  They're not strictly needed, but I think it reads more easily.  i.e `... in_progress = (activity == ZPOOL_WAIT_INITIALIZE) :`.

File: cmd/zpool/zpool_main.c (hunk 1 above)
Comment (ahrens) [reply cmd/zpool/zpool_main.c]: @behlendorf Do you know why we should not use `ASSERT`?  Should we make it not compile?

Comment (behlendorf) [reply cmd/zpool/zpool_main.c]: > Should we make it not compile?

Yes, we probably should.  From what I've seen the original code was quite good about using `assert(3)` for the userspace libraries and binaries.  As I understand it the ASSERT/VERIFY macros we provided as compatibility wrappers on all platforms solely to allow building the kernel code in user space (`libzpool`, etc).

Using them elsewhere doesn't lead to any problems, but it does muddle the long standing layering and can be confusing when looking at a source file which uses both.  If you look at `libzfs` you can see they've been slowing sneaking in over time.

**Pull request merged**